webapp/src/js/services/telemetry.js

```diff
@@ -0,0 +1,215 @@
+        $window.localStorage.setItem(DB_ID_KEY, dbName);
+      }
+      return $window.PouchDB(dbName); // avoid angular-pouch as digest isn't necessary here
+    };
+
+    var getUniqueDeviceId = function() {
+      var uniqueDeviceId = $window.localStorage.getItem(DEVICE_ID_KEY);
+      if (!uniqueDeviceId) {
+        uniqueDeviceId = uuid();
+        $window.localStorage.setItem(DEVICE_ID_KEY, uniqueDeviceId);
+      }
+
+      return uniqueDeviceId;
+    };
+
+    var getLastAggregatedDate = function() {
+      var date = parseInt(
+        $window.localStorage.getItem(LAST_AGGREGATED_DATE_KEY)
+      );
+      if (!date) {
+        date = Date.now();
+        $window.localStorage.setItem(LAST_AGGREGATED_DATE_KEY, date);
+      }
+      return date;
+    };
+
+    var storeIt = function(db, key, value) {
+      return db.post({
+        key: key,
+        value: value,
+        date_recorded: Date.now(),
+      });
+    };
+
+    var submitIfNeeded = function(db) {
+      var monthStart = moment().startOf('month');
+      var dbDate = moment(getLastAggregatedDate());
+      if (dbDate.isBefore(monthStart)) {
+        return aggregate(db).then(function() {
+          return reset(db);
+        });
+      }
+    };
+
+    var convertReduceToKeyValues = function(reduce) {
+      var kv = {};
+      reduce.rows.forEach(function(row) {
+        kv[row.key] = row.value;
+      });
+
+      return kv;
+    };
+
+    var generateAggregateDocId = function(metadata) {
+      return [
+        'telemetry',
+        metadata.year,
+        metadata.month,
+        metadata.user,
+        metadata.deviceId,
+      ].join('-');
+    };
+
+    var generateMetadataSection = function() {
+      var date = moment(getLastAggregatedDate());
+      return {
+        year: date.year(),
+        month: date.month(),
+        user: Session.userCtx().name,
+        deviceId: getUniqueDeviceId(),
+      };
+    };
+
+    var generateDeviceStats = function() {
+      return {
+        userAgent: $window.navigator.userAgent,
+        hardwareConcurrency: $window.navigator.hardwareConcurrency,
+        screen: {
+          width: $window.screen.availWidth,
+          height: $window.screen.availHeight,
+        },
+        // TODO: expose some device information in the medic-android wrapper
+        // and pull it in here. Device memory, phone type, stuff like that
+        //
+        // https://github.com/medic/medic-webapp/issues/4882
+      };
+    };
+
+    // This should never happen (famous last words..), because we should only
+    // generate a new document for every month, which is part of the _id.
+    var storeConflictedAggregate = function(aggregateDoc) {
+      aggregateDoc.metadata.conflicted = true;
+      aggregateDoc._id = [aggregateDoc._id, 'conflicted', Date.now()].join('-');
+      return DB().put(aggregateDoc);
+    };
+
+    var aggregate = function(db) {
+      return $q
+        .all([
+          db.query(
```

Comment:
Are there benefits or disadvantages of using a function here instead of a named view?

Reply:
I don't think so. The core reason it's inline here is that because you only ever call it once before destroying the DB, I didn't see the need to create a design document just to query it and then delete it.